include/zephyr/toolchain/common.h

```diff
@@ -175,6 +175,10 @@
 	}
 #endif
 
+// Same as BUILD_ASSERT(), but evaluates to an expression of value 0. Useful in macros.
+#define BUILD_ASSERT_EXPR(EXPR, MSG...) \
+	(0 * sizeof(struct { BUILD_ASSERT(EXPR, MSG); int dummy; }))
```

Comment:
As mentioned elsewhere, it's probably even more useful for this to take a parameter to which it will evaluate if it doesn't blow up.